numba/np/ufunc/deviceufunc.py

```diff
@@ -732,110 +724,180 @@ def _broadcast_scalar_input(self, ary, shape):
         raise NotImplementedError
 
 
-class GUFuncCallSteps(object):
+class GUFuncCallSteps(metaclass=ABCMeta):
+    """
+    Implements memory management and kernel launch operations for GUFunc calls.
+
+    One instance of this class is instantiated for each call, and the instance
+    is specific to the arguments given to the GUFunc call.
+
+    The base class implements the overall logic; subclasses provide
+    target-specific implementations of individual functions.
+    """
+
+    # The base class uses these slots; subclasses may provide additional slots.
     __slots__ = [
-        'args',
-        'kwargs',
-        'output',
-        'norm_inputs',
-        'kernel_returnvalue',
-        'kernel_parameters',
-        '_is_device_array',
-        '_need_device_conversion',
+        'outputs',
+        'inputs',
+        '_copy_result_to_host',
     ]
 
+    @abstractmethod
+    def launch_kernel(self, kernel, nelem, args):
+        """Implement the kernel launch"""
+
+    @abstractmethod
+    def is_device_array(self, obj):
+        """
+        Return True if `obj` is a device array for this target, False
+        otherwise.
+        """
+
+    @abstractmethod
+    def as_device_array(self, obj):
+        """
+        Return `obj` as a device array on this target.
+
+        May return `obj` directly if it is already on the target.
+        """
+
+    @abstractmethod
+    def to_device(self, hostary):
+        """
+        Copy `hostary` to the device and return the device array.
+        """
+
+    @abstractmethod
+    def allocate_device_array(self, shape, dtype):
+        """
+        Allocate a new uninitialized device array with the given shape and
+        dtype.
+        """
+
     def __init__(self, nin, nout, args, kwargs):
-        if nout > 1:
-            raise ValueError('multiple output is not supported')
-        self.args = args
-        self.kwargs = kwargs
-
-        user_output_is_device = False
-        self.output = self.kwargs.get('out')
-        if self.output is not None:
-            user_output_is_device = self.is_device_array(self.output)
-            if user_output_is_device:
-                self.output = self.as_device_array(self.output)
-        self._is_device_array = [self.is_device_array(a) for a in self.args]
-        self._need_device_conversion = (not any(self._is_device_array) and
-                                        not user_output_is_device)
-
-        # Normalize inputs
-        inputs = []
-        for a, isdev in zip(self.args, self._is_device_array):
-            if isdev:
-                inputs.append(self.as_device_array(a))
+        outputs = kwargs.get('out')
+
+        # Ensure the user has passed a correct number of arguments
+        if outputs is None and len(args) not in (nin, (nin + nout)):
+            def pos_argn(n):
+                s = f'{n} positional argument'
+                if n != 1:
+                    s += 's'
+                return s
```

Comment:
```suggestion
                return f'{n} positional argument{"s" * (n > 1)}'
```

Reply:
I think that needs to be `n != 1` to stop it saying "0 positional argument"

Reply:
Good point, suggestion adjusted.